internal/configs/node/config.go

```diff
@@ -251,37 +252,45 @@ func SetShardingSchedule(schedule shardingconfig.Schedule) {
 	}
 }
 
-// ShardIDFromConsensusKey returns the shard ID statically determined from the
-// consensus key.
-func (conf *ConfigType) ShardIDFromConsensusKey() (uint32, error) {
+// ShardIDFromKey returns the shard ID statically determined from the input key
+func (conf *ConfigType) ShardIDFromKey(key *bls.PublicKey) (uint32, error) {
 	var pubKey shard.BLSPublicKey
-	// all keys belong to same shard
-	if err := pubKey.FromLibBLSPublicKey(conf.ConsensusPubKey.PublicKey[0]); err != nil {
+	if err := pubKey.FromLibBLSPublicKey(key); err != nil {
 		return 0, errors.Wrapf(err,
 			"cannot convert libbls public key %s to internal form",
-			conf.ConsensusPubKey.SerializeToHexStr())
+			key.SerializeToHexStr())
 	}
 	epoch := conf.networkType.ChainConfig().StakingEpoch
 	numShards := conf.shardingSchedule.InstanceForEpoch(epoch).NumShards()
 	shardID := new(big.Int).Mod(pubKey.Big(), big.NewInt(int64(numShards)))
 	return uint32(shardID.Uint64()), nil
 }
 
+// ShardIDFromConsensusKey returns the shard ID statically determined from the
+// consensus key.
+func (conf *ConfigType) ShardIDFromConsensusKey() (uint32, error) {
+	return conf.ShardIDFromKey(conf.ConsensusPubKey.PublicKey[0])
```

Comment:
this [0] will never explode?

Reply:
nope. at least one bls key is always provided, including explorer for which default dummy bls is used.